### PR TITLE
drivers/ads101x: support for ADCs

### PR DESCRIPTION
**Pull request by ZetaR60**
Still a bit rough around the edges, but this adds support for the ADS101X and ADS111X ADCs. These are I2C 12-bit and 16-bit delta-sigma converters with optional programmable gain amp, input mux, and alarm comparator. Code is derived from drivers/adcxx1c.

Specific parts:
- ADS1013: 12-bit differential
- ADS1014: 12-bit differential with PGA and comparator
- ADS1015: 12-bit single or differential with mux, PGA, comparator (all testing on this part)
- ADS1113: 16-bit differential
- ADS1114: 16-bit differential with PGA and comparator
- ADS1115: 16-bit single or differential with mux, PGA, comparator

TODO:
- Alarm testing
- SAUL testing
- Only low speed I2C seems to be working, for some reason (should support fast)

### CODE REVIEW

**Comment by ZetaR60** [comment]
SAUL now tested and working.

**Comment by ZetaR60** [comment]
Alerts are now working. The I2C speed issues were caused by a second device on the bus being flaky.

Everything seems to be working properly.

File: tests/driver_ads101x/Makefile [hunk 1]
```diff
@@ -0,0 +1,15 @@
+include ../Makefile.tests_common
+
+FEATURES_REQUIRED = periph_i2c
+
+USEMODULE += ads101x
+
+# set alert parameters in case they are undefined
+TEST_ADS101X_LOW_LIMIT     ?= 10000
```
Comment (aabadie) [inline tests/driver_ads101x/Makefile]: Why not directly use those values in the ads101x_params.h driver file ?

Comment (ZetaR60) [reply tests/driver_ads101x/Makefile]: Good question. The answer is... no reason except that the adcxx1c driver did it that way. I have put them in ads101x_params.h

File: drivers/ads101x/include/ads101x_params.h [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @name    Set default configuration parameters for the ADS101x/111x driver
+ * @{
+ */
+#ifndef ADS101X_PARAM_NAME
```
Comment (aabadie) [inline drivers/ads101x/include/ads101x_params.h]: This parameter is unused, I would remove it

Comment (aabadie) [reply drivers/ads101x/include/ads101x_params.h]: ok, but then you should this kind of struct:
```c
static const saul_reg_info_t ads101s_saul_info[] =
{
    ADS101X_SAUL_INFO
};
```
like it's done for (nearly) all other drivers

Comment (ZetaR60) [reply drivers/ads101x/include/ads101x_params.h]: I don't know why I did it that way. Fixed.

**Comment by aabadie** [comment]
I don't have the hardware for testing but could you just post an output of the running test application ? Or maybe someone else could try ?

**Comment by ZetaR60** [comment]
```
Installed compiler toolchains 
-----------------------------
           native gcc: missing
    arm-none-eabi-gcc: missing
              avr-gcc: avr-gcc (Fedora 7.2.0-1.fc27) 7.2.0
     mips-mti-elf-gcc: missing
           msp430-gcc: missing
                clang: missing

Installed compiler libs
-----------------------
 arm-none-eabi-newlib: missing
  mips-mti-elf-newlib: missing
             avr-libc: "2.0.0" ("20150208")

Installed development tools
---------------------------
                cmake: cmake version 3.11.0
             cppcheck: missing
              doxygen: missing
               flake8: missing
                  git: git version 2.14.2
           coccinelle: missing
```

Board: mega-xplained


Channel 1 has a NTC thermistor that I put my finger on, which makes the reading drop.
Output of tests/driver_ads101x:
```
/home/liveuser/Desktop/RIOT_ads101x/dist/tools/pyterm/pyterm -p "/dev/ttyUSB0" -b "9600"
No handlers could be found for logger "root"
2018-04-13 11:15:12,869 - INFO # Connect to serial port /dev/ttyUSB0
Welcome to pyterm!
Type '/exit' to exit.
2018-04-13 11:15:13,881 - INFO #  17664

2018-04-13 11:15:13,887 - INFO # main(): This is RIOT! (Version: 2018.04-devel-827-g62cfc-localhost-live-RIOT_ads101x)
2018-04-13 11:15:13,889 - INFO # ADS101X analog to digital driver test application
2018-04-13 11:15:13,889 - INFO # 
2018-04-13 11:15:13,890 - INFO # Initializing ADS101x analog to digital at I2C_DEV(0)... [OK]
2018-04-13 11:15:13,892 - INFO # Initializing ADS101x alert at I2C_DEV(0)... [OK]
2018-04-13 11:15:13,893 - INFO # Enabling alert interrupt: [OK]
2018-04-13 11:15:13,894 - INFO # Raw analog read. CH0: 5344 CH1: 17616 CH2: 32752 CH3: 32752
2018-04-13 11:15:13,895 - INFO # Raw analog read. CH0: 5520 CH1: 17600 CH2: 32752 CH3: 32752
2018-04-13 11:15:13,897 - INFO # Raw analog read. CH0: 5456 CH1: 17584 CH2: 32752 CH3: 32752
2018-04-13 11:15:13,898 - INFO # Raw analog read. CH0: 5488 CH1: 17552 CH2: 32752 CH3: 32752
2018-04-13 11:15:13,899 - INFO # Raw analog read. CH0: 5568 CH1: 17552 CH2: 32752 CH3: 32752
2018-04-13 11:15:13,926 - INFO # Raw analog read. CH0: 5520 CH1: 17520 CH2: 32752 CH3: 32752
2018-04-13 11:15:14,054 - INFO # Raw analog read. CH0: 5424 CH1: 17504 CH2: 32752 CH3: 32752
2018-04-13 11:15:14,182 - INFO # Raw analog read. CH0: 5616 CH1: 17488 CH2: 32752 CH3: 32752
2018-04-13 11:15:14,294 - INFO # Raw analog read. CH0: 5568 CH1: 17472 CH2: 32752 CH3: 32752
2018-04-13 11:15:14,422 - INFO # Raw analog read. CH0: 5520 CH1: 17440 CH2: 32752 CH3: 32752
2018-04-13 11:15:14,534 - INFO # Raw analog read. CH0: 5376 CH1: 17424 CH2: 32752 CH3: 32752
2018-04-13 11:15:14,662 - INFO # Raw analog read. CH0: 5472 CH1: 17408 CH2: 32752 CH3: 32752
2018-04-13 11:15:14,789 - INFO # Raw analog read. CH0: 5568 CH1: 17376 CH2: 32752 CH3: 32752
2018-04-13 11:15:14,901 - INFO # Raw analog read. CH0: 5552 CH1: 17360 CH2: 32752 CH3: 32752
2018-04-13 11:15:15,029 - INFO # Raw analog read. CH0: 5632 CH1: 17344 CH2: 32752 CH3: 32752
2018-04-13 11:15:15,141 - INFO # Raw analog read. CH0: 5488 CH1: 17328 CH2: 32752 CH3: 32752
2018-04-13 11:15:15,269 - INFO # Raw analog read. CH0: 5424 CH1: 17296 CH2: 32752 CH3: 32752
2018-04-13 11:15:15,397 - INFO # Raw analog read. CH0: 5392 CH1: 17280 CH2: 32752 CH3: 32752
2018-04-13 11:15:15,509 - INFO # Raw analog read. CH0: 5600 CH1: 17264 CH2: 32752 CH3: 32752
2018-04-13 11:15:15,637 - INFO # Raw analog read. CH0: 5488 CH1: 17232 CH2: 32752 CH3: 32752
2018-04-13 11:15:15,749 - INFO # Raw analog read. CH0: 5440 CH1: 17232 CH2: 32752 CH3: 32752
2018-04-13 11:15:15,877 - INFO # Raw analog read. CH0: 5472 CH1: 17216 CH2: 32752 CH3: 32752
2018-04-13 11:15:16,005 - INFO # Raw analog read. CH0: 5488 CH1: 17184 CH2: 32752 CH3: 32752
2018-04-13 11:15:16,117 - INFO # Raw analog read. CH0: 5552 CH1: 17168 CH2: 32752 CH3: 32752
2018-04-13 11:15:16,245 - INFO # Raw analog read. CH0: 5472 CH1: 17152 CH2: 32752 CH3: 32752
2018-04-13 11:15:16,357 - INFO # Raw analog read. CH0: 5568 CH1: 17136 CH2: 32752 CH3: 32752
2018-04-13 11:15:16,485 - INFO # Raw analog read. CH0: 5552 CH1: 17120 CH2: 32752 CH3: 32752
2018-04-13 11:15:16,597 - INFO # Raw analog read. CH0: 5504 CH1: 17088 CH2: 32752 CH3: 32752
2018-04-13 11:15:16,725 - INFO # Raw analog read. CH0: 5552 CH1: 17072 CH2: 32752 CH3: 32752
2018-04-13 11:15:16,853 - INFO # Raw analog read. CH0: 5504 CH1: 17056 CH2: 32752 CH3: 32752
2018-04-13 11:15:16,965 - INFO # Raw analog read. CH0: 5520 CH1: 17040 CH2: 32752 CH3: 32752
2018-04-13 11:15:17,047 - INFO # Exiting Pyterm
```

For SAUL I added the following:
boards/mega-xplained/Makefile.dep:
```
USEMODULE += boards_common_atmega

ifneq (,$(filter saul_default,$(USEMODULE)))
  USEMODULE += ads101x
endif
```

boards/mega-xplained/include/ads101x_params.h:
```
#include "board.h"
#include "saul_reg.h"
#include "ads101x.h"
#include "ads101x_regs.h"

static const ads101x_params_t ads101x_params[] =
{
    {
     	.i2c = I2C_DEV(0),
        .addr = ADS101X_I2C_ADDRESS,
        .mux_gain  = (ADS101X_AIN0_SINGM | ADS101X_PGA_FSR_2V048)
    },
    {
     	.i2c = I2C_DEV(0),
        .addr = ADS101X_I2C_ADDRESS,
        .mux_gain  = (ADS101X_AIN1_SINGM | ADS101X_PGA_FSR_2V048)
    },
    {
     	.i2c = I2C_DEV(0),
        .addr = ADS101X_I2C_ADDRESS,
        .mux_gain  = (ADS101X_AIN2_SINGM | ADS101X_PGA_FSR_2V048)
    },
    {
     	.i2c = I2C_DEV(0),
        .addr = ADS101X_I2C_ADDRESS,
        .mux_gain  = (ADS101X_AIN3_SINGM | ADS101X_PGA_FSR_2V048)
    }
};

static const saul_reg_info_t ads101x_saul_info[] =
{
    { .name = "CH0" },
    { .name = "CH1" },
    { .name = "CH2" },
    { .name = "CH3" }
};
```

Output of examples/saul:
```
/home/liveuser/Desktop/RIOT_ads101x/dist/tools/pyterm/pyterm -p "/dev/ttyUSB0" -b "9600"
No handlers could be found for logger "root"
2018-04-13 11:30:05,394 - INFO # Connect to serial port /dev/ttyUSB0
Welcome to pyterm!
Type '/exit' to exit.
2018-04-13 11:30:06,405 - INFO # 

2018-04-13 11:30:06,412 - INFO # main(): This is RIOT! (Version: 2018.04-devel-827-g62cfc-localhost-live-RIOT_ads101x)
2018-04-13 11:30:06,414 - INFO # Welcome to RIOT!
2018-04-13 11:30:06,415 - INFO # 
2018-04-13 11:30:06,416 - INFO # Type `help` for help, type `saul` to see all SAUL devices
2018-04-13 11:30:06,416 - INFO # 
> help
2018-04-13 11:30:08,948 - INFO #  help
2018-04-13 11:30:08,979 - INFO # Command              Description
2018-04-13 11:30:09,027 - INFO # ---------------------------------------
2018-04-13 11:30:09,059 - INFO # reboot               Reboot the node
2018-04-13 11:30:09,123 - INFO # ps                   Prints information about running threads.
2018-04-13 11:30:09,188 - INFO # saul                 interact with sensors and actuators using SAUL
> saul
2018-04-13 11:30:10,436 - INFO #  saul
2018-04-13 11:30:10,452 - INFO # ID	Class		Name
2018-04-13 11:30:10,467 - INFO # #0	SENSE_ANALOG	CH0
2018-04-13 11:30:10,499 - INFO # #1	SENSE_ANALOG	CH1
2018-04-13 11:30:10,515 - INFO # #2	SENSE_ANALOG	CH2
2018-04-13 11:30:10,532 - INFO # #3	SENSE_ANALOG	CH3
> saul read 0
2018-04-13 11:30:13,635 - INFO #  saul read 0
2018-04-13 11:30:13,682 - INFO # Reading from #0 (CH0|SENSE_ANALOG)
2018-04-13 11:30:13,698 - INFO # Data:	     0.3360V
> saul read 1
2018-04-13 11:30:16,211 - INFO #  saul read 1
2018-04-13 11:30:16,258 - INFO # Reading from #1 (CH1|SENSE_ANALOG)
2018-04-13 11:30:16,274 - INFO # Data:	     1.1430V
> saul read 2
2018-04-13 11:30:18,978 - INFO #  saul read 2
2018-04-13 11:30:19,025 - INFO # Reading from #2 (CH2|SENSE_ANALOG)
2018-04-13 11:30:19,042 - INFO # Data:	     2.0470V
> saul read 3
2018-04-13 11:30:21,666 - INFO #  saul read 3
2018-04-13 11:30:21,714 - INFO # Reading from #3 (CH3|SENSE_ANALOG)
2018-04-13 11:30:21,730 - INFO # Data:	     2.0470V
> 2018-04-13 11:30:25,308 - INFO # Exiting Pyterm
```

**Comment by ZetaR60** [comment]
Squashed. Thanks for the help aabadie!

**Comment by ZetaR60** [comment]
ping @aabadie

**Pull request merged**